core/src/main/java/org/elasticsearch/transport/netty/NettyTransport.java

```diff
@@ -485,48 +495,61 @@ public boolean onPortNumber(int portNumber) {
-                // other profiles together, we need to change how the profileSettings are built for the default profile...
-                int publishPort = settings.getAsInt("transport.netty.publish_port", settings.getAsInt("transport.publish_port", boundAddress.getPort()));
-                String publishHosts[] = settings.getAsArray("transport.netty.publish_host", settings.getAsArray("transport.publish_host", settings.getAsArray("transport.host", null)));
-                InetSocketAddress publishAddress = createPublishAddress(publishHosts, publishPort);
-                this.boundAddress = new BoundTransportAddress(new TransportAddress[]{new InetSocketTransportAddress(boundAddress)}, new InetSocketTransportAddress(publishAddress));
-            } else {
-                // the default profile is already bound to one address and has the publish address, copy the existing bound addresses as is and append the new address.
-                // the new bound address is appended in order to preserve the ordering/priority of bound addresses
-                TransportAddress[] existingBoundAddress = this.boundAddress.boundAddresses();
-                TransportAddress[] updatedBoundAddresses = Arrays.copyOf(existingBoundAddress, existingBoundAddress.length + 1);
-                updatedBoundAddresses[updatedBoundAddresses.length - 1] = new InetSocketTransportAddress(boundAddress);
-                this.boundAddress = new BoundTransportAddress(updatedBoundAddresses, this.boundAddress.publishAddress());
+            publishHosts = profileSettings.getAsArray("publish_host", boundAddressesHostStrings);
+        }
+
+        final InetAddress publishInetAddress;
+        try {
+            publishInetAddress = networkService.resolvePublishHostAddresses(publishHosts);
+        } catch (Exception e) {
+            throw new BindTransportException("Failed to resolve publish address", e);
+        }
+
+        Integer publishPort;
+        if (DEFAULT_PROFILE.equals(name)) {
+            publishPort = settings.getAsInt("transport.netty.publish_port", settings.getAsInt("transport.publish_port", null));
+        } else {
+            publishPort = profileSettings.getAsInt("publish_port", null);
+        }
+
+        // if port not explicitly provided, search for port of address in boundAddresses that matches publishInetAddress
+        if (publishPort == null) {
+            for (InetSocketAddress boundAddress : boundAddresses) {
+                if (boundAddress.getAddress().equals(publishInetAddress)) {
+                    publishPort = boundAddress.getPort();
+                    break;
+                }
             }
         }
 
-        if (logger.isDebugEnabled()) {
-            logger.debug("Bound profile [{}] to address {{}}", name, NetworkAddress.format(boundSocket.get()));
+        // if port still not matches, just take port of first bound address
+        if (publishPort == null) {
```

Comment:
If its unspecified, and we automatically created the publish address, then the only situation where we should not find a matching publish address should be wildcard bind addresses (`0.0.0.0`, `::`). In this case we take the wildcard bind address and "expand" the wildcard to select a publish address (https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/common/network/NetworkService.java#L172-L176)

I'm not sure if its easy to add an assert (given transport profiles) but maybe a code comment for the future. It would be nice somewhere in the future to check this is actually the case and throw exception otherwise.